hoppscotch/hoppscotch
# feat: description field for request parameters and headers

### Description

This PR introduces a new `description` field for both request parameters and headers in the API documentation. This allows developers to provide clear and concise explanations for each parameter and header, enhancing the readability and usability of the documentation. It is also added for headers at the collection level.

The `description` field is also accounted for while importing from external sources (`Postman`, `Insomnia` & `OpenAPI`).

A pre-existing issue where edits made to headers from the bulk edit context are not immediately reflected in the default view is resolved.

Closes HFE-535 HFE-538.

### Preview

![image](https://github.com/user-attachments/assets/d4b2edde-486b-4c83-ba8e-003049494ea2)

> For the auto-generated `Authorization` and `Content-Type` headers, the `description` field will stay disabled and the value will appear empty; while for inherited headers the value specified will be populated.

![header-desc](https://github.com/user-attachments/assets/1a40e2dc-040c-4759-8da0-7e94d5ca9457)

### Changes

- Adds a new `HttpKeyValue` component abstracting the active headers UI and related business logic for reuse across request `Parameters` & `Headers` tabs.
- Update bulk edit context and the default view synchronization logic accounting for the newly added `description` field. The `description` field is excluded while comparing against the key-value pairs from the bulk edit context since it isn't applicable.
- Changes to versioned entities:
  - `HoppCollection` schema `v3` - Headers at the collection level have an additional `description` field. Also, the `v2` schema is updated to consume `HoppRESTHeaders` & `GQLHeader` schema definitions without the `description` field addition. Also, `HoppRESTAuth` & `HoppGQLAuth` schema definitions are consumed from the `HoppRESTRequest` & `HoppGQLRequest` `v5` schemas instead of from the entry point. This is done so that any field additions at the child level (`auth`, `headers`, etc) will require adding a new version for `HoppCollection` to stay consistent and won't be implicitly supported by the existing latest version as was the case.

    > Reasoning: Consider the pre-existing scenario where a new field is added under headers `HoppRESTHeaders`. Assume `HoppCollection` `v2` points to the entry point where the latest version of the above is exported. For an incoming collection conforming to the `v2` schema, there is no proper migration mechanism accounting for the `description` field addition and making it point to the specific `HoppRESTHeaders` schema and introducing `HoppCollection` `v3` with the relevant migration mechanism adding `description` field falling back to the value of an empty string ensures proper migration mechanism comes into play and data conforming to the latest version is dealt with.

  - `HoppRESTRequest` schema `v7` - New `description` field for headers and parameters.
  - `HoppGQLRequest` schema `v6` - New `description` field for headers.
- Importer module updates to include the `description` field from external sources.
- Updates to test suites and associated fixtures accounting for the internal data structure updates.
- Relevant `i18n` string additions.
- Update `Node.js CI` GH Action workflow config to include `next` & `patch` branches as triggers on `push` & `pull_request` events.

### Note to reviewers

Ensure to verify the following flows:

- Specify a description under request parameters/headers for an unsaved request and ensure it persists across page reloads.
- Save the request (`REST`/`GQL`) under a collection and observe the behaviour. Try exporting (individual/workspace level exports) and importing the same and verify if the `description` fields are populated. This applies to both personal and team workspaces.
- Navigate to the `Authorization` tab, change the type and observe the `Authorization` header addition with the `description` field appearing empty and staying disabled.
- Navigate to the bulk edit view and verify edits are synced to and fro.
- Open the collection properties modal, specify a description and ensure it persists. For a request inheriting the authorization/header properties, ensure the `description` field gets populated. Try changing the value at the collection level and verify it gets reflected. Inherited headers will stay disabled at a request level.
- Import Postman, Insomnia & OpenAPI collections with descriptions set for request parameters/headers and observe them getting populated while importing.

We're following the convention where any changes under `@hoppscotch/data` will require a CLI version bump (`@hoppscotch/data` contents get inlined in the CLI bundle), especially in this case where a new schema definition is added and the CLI would reject a collection conforming to the latest version without publishing the latest changes. The bump is done in #4275 scheduled for the next major release, hence not done in this PR.
